### ISSUE TITLE
Question about punycode/internationalized domains

### ISSUE THREAD

**Issue by nonk123** [caddyserver/caddy#6404]
Spent 3 days banging my head on the wall while figuring this out, but the solution was so simple.

I tried this in my `Caddyfile`:

```
https://подработка-павлово.рф {
    reverse_proxy docker_container:8000
}
```

This always got me a 0-byte response whenever I opened the link in my browser.

Then I changed it to punycode:

```
https://xn----7sbabcjaj0doydgbbik3b.xn--p1ai {
    reverse_proxy docker_container:8000
}
```

And voila! Everything works.

Was I missing something? I haven't seen any mention of internationalized domains being unsupported in the docs.

Perhaps it should be pointed out more clearly? Or even better: support these by default?

**Comment by mholt** [comment]
They should be supported. Let's fix this.

**Comment by mholt** [comment]
Ok now I remember adding [support for this in CertMagic](https://github.com/caddyserver/certmagic/blob/193db7523a09302e32772263dfe050f7dd9c20a9/config.go#L997-L1002
) a long time ago:

```go
// convert IDNs to ASCII according to RFC 5280 section 7
normalizedName, err := idna.ToASCII(name)
if err != nil {
	return nil, fmt.Errorf("converting identifier '%s' to ASCII: %v", name, err)
}
csrTemplate.DNSNames = append(csrTemplate.DNSNames, normalizedName)
```

So off-hand, I am not sure why this doesn't work. :thinking: Will need to dig more.

**Comment by mholt** [comment]
I remembered to handle this for CertMagic, so that's fine; but the `host` matcher does a literal string comparison. We probably need to convert IDNs to their punycode equivalents since that's what clients send even if you input an IDN to them.

**Comment by mholt** [comment]
Fixed in 99dcdf7e426f0dcbdffe510f241ae8a4fd5a56e6 -- kind of surprised this hasn't come up before. :man_shrugging:  Thanks for the issue!
